### PR TITLE
PP-4779: Add/Replace gateway merchant id successfully

### PR DESCRIPTION
**Pull request by oswaldquek**
We could have used the `PATCH /v1/frontend/accounts/{accountId}` method to
update the credentials but operationally curling `PATCH
/v1/frontend/accounts/{accountId}/gateway-merchant-id/{gatewayMerchantId}` manually is
simpler and will be deprecated once the functionality in self service to update
a gateway account with the gateway merchant id is in place.

@oswaldquek @gidsg

### CODE REVIEW

File: src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountFrontendResourceITest.java [hunk 1]
```diff
@@ -139,6 +139,21 @@ public void updateAndGetGatewayAccountWithWorldpayMerchantId() {
                 .then()
                 .statusCode(HttpStatus.SC_OK)
                 .body("gateway_merchant_id", is("1234abc"));
+
+        givenSetup().accept(JSON)
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountFrontendResourceITest.java]: @gidsg I think this block can be deleted?

File: src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java [hunk 2]
```diff
@@ -238,6 +238,31 @@ public Response patchGatewayAccount(@PathParam("accountId") Long gatewayAccountI
                 .orElseGet(() -> Response.status(NOT_FOUND).build());
     }
 
+    //Deprecate when self service has functionality to update the gateway merchant id
+    @PATCH
+    @Path("/v1/frontend/accounts/{accountId}/gateway-merchant-id/{gatewayMerchantId}")
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java]: Can we post 'gatewayMerchantId' in body rather than pass in URL ?
Requests are logged and we will log merchant Id if it is passed in URL.

File: src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java [hunk 3]
```diff
@@ -238,6 +238,33 @@ public Response patchGatewayAccount(@PathParam("accountId") Long gatewayAccountI
...
+    @Consumes(APPLICATION_JSON)
+    @Transactional
+    public Response updateGatewayAccountCredentialsWithGatewayMerchantId(@PathParam("accountId") Long gatewayAccountId, @PathParam("gatewayMerchantId") String gatewayMerchantId) {
+        return gatewayDao.findById(gatewayAccountId).map(gatewayAccount -> {
+            Map<String, String> credentials = gatewayAccount.getCredentials();
+            Map<String, String> updatedCredentials = new HashMap<>(credentials);
+            updatedCredentials.put("gateway_merchant_id", gatewayMerchantId);
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java]: > Map<String, String> credentials = gatewayAccount.getCredentials();
            Map<String, String> updatedCredentials = new HashMap<>(credentials);
            updatedCredentials.put("gateway_merchant_id", gatewayMerchantId);

Is there any reason a new HashMap is created here instead of something like
```
gatewayAccount
.getCredentials()
.put("gateway_merchant_id", gatewayMerchantId);
```

File: src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java [hunk 4]
```diff
@@ -238,6 +238,33 @@ public Response patchGatewayAccount(@PathParam("accountId") Long gatewayAccountI
                 .orElseGet(() -> Response.status(NOT_FOUND).build());
     }
 
+    //Deprecate when self service has functionality to update the gateway merchant id
+    @PATCH
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java]: I suppose there is going to be some functionality in toolbox to use this endpoint ?

File: src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java [hunk 5]
```diff
@@ -238,6 +238,33 @@ public Response patchGatewayAccount(@PathParam("accountId") Long gatewayAccountI
                 .orElseGet(() -> Response.status(NOT_FOUND).build());
     }
 
+    //Deprecate when self service has functionality to update the gateway merchant id
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java]: Is there any story to update self service ? Would be good to add reference here so someone can find and deprecate/remove this endpoint

File: src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java [hunk 6]
```diff
@@ -238,6 +238,33 @@ public Response patchGatewayAccount(@PathParam("accountId") Long gatewayAccountI
...
+            gatewayAccount.setCredentials(updatedCredentials);
+            return Response.ok().build();
+        }).orElseGet(() -> notFoundResponse(format("The gateway account id '%s' does not exist", gatewayAccountId)));
+    }
+
+    @GET
+    @Path("/v1/frontend/accounts/{accountId}/gateway-merchant-id")
+    @Produces(APPLICATION_JSON)
+    public Response getGatewayMerchantId(@PathParam("accountId") Long gatewayAccountId) {
+        return gatewayDao.findById(gatewayAccountId).map(gatewayAccount -> {
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java]: This went back to old ways of connector code with all functionality (transactional boundary, db updates, error handling.. ) in this endpoint. Given that we may remove endpoint (and with all other endpoints in this resource following same approach), don't see any issue to keep this way. But if this is going to be stay for a while we will need to refactor code into appropriate service.

File: src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountFrontendResourceITest.java [hunk 7]
```diff
@@ -121,6 +122,51 @@ String getMerchantId() {
...
+
+        givenSetup()
+                .get(ACCOUNTS_FRONTEND_URL + accountId + "/gateway-merchant-id")
+                .then()
+                .statusCode(HttpStatus.SC_OK)
+                .body("gateway_merchant_id", is("1234abc"));
+    }
+
+    @Test
+    public void getGatewayAccountWithInvalidAccount(){
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountFrontendResourceITest.java]: Can we name it similar to other tests method names `shouldReturn404IfGatewayAccountDoesNotExist` and to be descriptive

File: src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountFrontendResourceITest.java [hunk 8]
```diff
@@ -121,6 +122,51 @@ String getMerchantId() {
...
+        givenSetup()
+                .get(ACCOUNTS_FRONTEND_URL + "99999999" + "/gateway-merchant-id")
+                .then()
+                .statusCode(HttpStatus.SC_NOT_FOUND)
+                .body("message", is("The gateway account id '99999999' does not exist"));
+    }
+
+    @Test
+    public void getGatewayAccountWithoutMerchantId(){
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountFrontendResourceITest.java]: Same as above for naming..

File: src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java (hunk 2 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java]: Path could be `/v1/accounts/{accountId}/gateway-merchant-id/{gatewayMerchantId}` if we are not going to invoke it through frontend/selfservice

**Review by kbottla — CHANGES_REQUESTED** [review]

**Comment by oswaldquek** [comment]
Do not merge. I think the requirements around this story need to be slightly updated.

File: src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java (hunk 3 above)
Comment (oswaldquek) [reply src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java]: yeah we tried that. it doesn't work. i think it's something to do with JPA seeing the same object (Map) reference and thinks it doesn't need to update the credentials.

File: src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java (hunk 6 above)
Comment (oswaldquek) [reply src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java]: I will remove this endpoint. I just spoke to Jon and it will be easier to return the gateway merchant id in the call to `/v1/frontend/charges/{chargeId}` so frontend doesn't have to make an extra api call.

File: src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java (hunk 2 above)
Comment (oswaldquek) [reply src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java]: yup all good points. will fix these.

File: src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java (hunk 4 above)
Comment (oswaldquek) [reply src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java]: toolbox and selfservice will use https://github.com/alphagov/pay-connector/blob/67e8a2378cb4152887e235b0b39f870fc5a45e53/src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java#L232

i created this method to make it easier for me to manually update the gateway account with the gateway merchant id using `curl`. this is only temporary and when selfservice is updated to allow services to to update their account with their gateway merchant id this endpoint will be deleted.

File: src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java (hunk 5 above)
Comment (oswaldquek) [reply src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java]: yes in here https://payments-platform.atlassian.net/browse/PP-4645. it's not prioritised yet though. there will probably be a few stories relating to self service.

File: src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java (hunk 2 above)
Comment (oswaldquek) [reply src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java]: i fixed this by deleting the path ;)
`PATCH /v1/api/accounts/accountId` will be used instead

File: src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java (hunk 4 above)
Comment (oswaldquek) [reply src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java]: fixed

File: src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java (hunk 5 above)
Comment (oswaldquek) [reply src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java]: fixed

File: src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountFrontendResourceITest.java (hunk 7 above)
Comment (oswaldquek) [reply src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountFrontendResourceITest.java]: this test is deleted

File: src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountFrontendResourceITest.java (hunk 8 above)
Comment (oswaldquek) [reply src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountFrontendResourceITest.java]: this test is deleted

File: src/main/java/uk/gov/pay/connector/gatewayaccount/model/GatewayAccountEntity.java [hunk 9]
```diff
@@ -170,6 +170,12 @@ public String getGatewayName() {
         return credentials;
     }
 
+    @JsonProperty("gateway_merchant_id")
+    @JsonView(Views.FrontendView.class)
+    public String getGatewayMerchantId() {
+        return credentials.get("gateway_merchant_id");
```
Comment (gidsg) [inline src/main/java/uk/gov/pay/connector/gatewayaccount/model/GatewayAccountEntity.java:176]: does this handle the possible null case?

Comment (oswaldquek) [reply src/main/java/uk/gov/pay/connector/gatewayaccount/model/GatewayAccountEntity.java:176]: it doesn't matter as we usually return "null" if there's nothing. for example, 
```
    "gateway_account": {
        "version": 3,
        "requires3ds": false,
        "allowWebPayments": false,
        "notifySettings": null,
        "live": false,
        "gateway_account_id": 1,
        "payment_provider": "worldpay",
        "type": "test",
        "service_name": null,
        "analytics_id": null,

```

File: src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountRequestValidator.java [hunk 10]
```diff
@@ -75,6 +81,11 @@ public void validatePatchRequest(JsonNode payload){
         }
     }
 
+    private void validateGatewayMerchantId(JsonNode payload) {
+        throwIfInvalidFieldOperation(payload, ADD_OP);
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountRequestValidator.java]: I think we need `replace` op too for updating existing

File: src/main/java/uk/gov/pay/connector/gatewayaccount/service/GatewayAccountService.java [hunk 11]
```diff
@@ -62,13 +63,20 @@ public GatewayAccountResponse createGatewayAccount(GatewayAccountRequest gateway
...
     }
 
 
     private final Map<String, BiConsumer<PatchRequest, GatewayAccountEntity>> attributeUpdater =
             new HashMap<String, BiConsumer<PatchRequest, GatewayAccountEntity>>() {{
+                put(CREDENTIALS_GATEWAY_MERCHANT_ID,
+                        (gatewayAccountRequest, gatewayAccountEntity) -> {
+                            Map<String, String> credentials = gatewayAccountEntity.getCredentials();
+                            Map<String, String> updatedCredentials = new HashMap<>(credentials);
+                            updatedCredentials.put("gateway_merchant_id", gatewayAccountRequest.valueAsString());
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/gatewayaccount/service/GatewayAccountService.java:76]: We need to check if gateway_merchant_id exists and update - if we allow `replace` op

File: src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountRequestValidator.java [hunk 12]
```diff
@@ -57,6 +60,9 @@ public void validatePatchRequest(JsonNode payload){
             throw new ValidationException(Collections.singletonList(format("Operation [%s] not supported for path [%s]", FIELD_OPERATION, path)));
         
         switch (path) {
+            case CREDENTIALS_GATEWAY_MERCHANT_ID:
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountRequestValidator.java:64]: I would like to see tests for this in `GatewayAccountRequestValidatorTest`

**Review by kbottla — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountRequestValidator.java (hunk 12 above)
Comment (oswaldquek) [reply src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountRequestValidator.java:64]: do we really need tests in `GatewayAccountRequestValidatorTest` as the positive test is covered by `GatewayAccountFrontendResourceITest. updateGatewayAccountWithGatewayMerchantId`? also, looking at `GatewayAccountRequestValidatorTest` there are only two positive tests `shouldNotThrow_whenAllValidationsPassed` and `shouldNotThrow_whenPathIsValidEmailCollectionMode` which suggests it's overkill to test everything at this unit test level as the rest are tested in the integration tests.

i'm also not convinced about testing for the negative as the validation of the gateway merchant id uses `throwIfInvalidFieldOperation` and `throwIfNullFieldValue` which are already tested.

Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountRequestValidator.java:64]: I am happy about the positive test as it is covered in Integration test but we don't cover negative cases anywhere.

Although `throwIfInvalidFieldOperation` and `throwIfNullFieldValue` are tested, the method `validateGatewayMerchantId` itself is not. It means we can change it to anything without detected.

File: src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountRequestValidator.java (hunk 10 above)
Comment (oswaldquek) [reply src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountRequestValidator.java]: I've created a test in `GatewayMerchantIdITest.replaceGatewayMerchantId`

File: src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountRequestValidator.java (hunk 12 above)
Comment (oswaldquek) [reply src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountRequestValidator.java:64]: i've created a negative test in `GatewayAccountRequestValidatorTest.shouldThrowForInvalidMerchantGatewayIdPatchRequest`

Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountRequestValidator.java:64]: 👍

**Review by kbottla — APPROVED** [review]
LGTM! You might want to update PR description since it doesn't reflect current state of changes
